tests/test_objective_funs.py

```diff
@@ -569,6 +576,176 @@ def test(eq):
+        nonaxisym_field = FourierCurrentPotentialField.from_surface(
+            surface=surf,
+            Phi_mn=np.array([1e4]),
+            modes_Phi=np.array([[1, 0]]),
+            G=1e4,
+            I=0,
+            sym_Phi="sin",
+        )
+
+        return nonaxisym_field
+
+    @pytest.mark.unit
+    def test_quadratic_flux(self, quadratic_flux_equilibriums):
+        """Test calculation of quadratic flux on the boundary."""
+        t_field = ToroidalMagneticField(1, 1)
+
+        # test that torus (axisymmetric) Bnorm is exactly 0
+        __, eq = quadratic_flux_equilibriums
+        obj = QuadraticFlux(t_field, eq)
+        obj.build(eq, verbose=2)
+        f = obj.compute(field_params=t_field.params_dict)
+        np.testing.assert_allclose(f, 0, rtol=1e-15, atol=1e-15)
+
+        # test nonaxisymmetric surface
+        eq = desc.io.load("desc/examples/precise_QA_output.h5")[0]
+
+        __, eval_grid, source_grid = self.get_grids(eq, t_field)
+
+        obj = QuadraticFlux(
+            t_field,
+            eq,
+            eval_grid=eval_grid,
+            src_grid=source_grid,
+        )
+        Bnorm = t_field.compute_Bnormal(
+            eq.surface, eval_grid=eval_grid, source_grid=source_grid
+        )[0]
+        obj.build(eq)
+        f = obj.compute(field_params=t_field.params_dict)
+
+        np.testing.assert_allclose(f, Bnorm, atol=1e-3)
+
+    @pytest.mark.unit
+    def test_quadratic_flux_with_eq_fixed(self, quadratic_flux_equilibriums):
+        """Test with eq_fixed = True, field_fixed = False.
+
+        Checks that a non-axisymmetric field becomes axisymmetric
+        given an axisymmetric field.
+        """
+        __, eq = quadratic_flux_equilibriums
+
+        # create a winding surface that surrounds the equilibrium
+        field = self.make_nonaxisym_field(eq)
+
+        optimizer = Optimizer("lsq-exact")
+
+        field_grid, eval_grid, source_grid = self.get_grids(eq, field)
+
+        # fix every parameter except for phi_mn
+        constraints = (FixParameter(field, ["I", "G", "R_lmn", "Z_lmn"]),)
+
+        quadflux_obj = QuadraticFlux(
+            ext_field=field,
+            eq=eq,
+            src_grid=source_grid,
+            eval_grid=eval_grid,
+            field_grid=field_grid,
+        )
+        objective = ObjectiveFunction(quadflux_obj)
+
+        things, __ = optimizer.optimize(
+            field,
+            objective=objective,
+            constraints=constraints,
+            copy=True,
+            ftol=1e-14,
+            gtol=1e-14,
+        )
+
+        np.testing.assert_allclose(things[0].Phi_mn, 0, atol=1e-8)
+
+    @pytest.mark.unit
+    def test_quadratic_flux_with_analytic_field(self):
+        """Test analytic field optimization when eq_fixed=True, field_fixed=False.
+
+        Checks that B goes to zero for axisymmetric eq and field.
+        """
+        eq = desc.examples.get("precise_QA")
```

Comment:
can reduce resolution a lot after loading to save time

`eq.change_resolution(3,3,3,6,6,6)` or something